### ISSUE TITLE
 "TypeError: list indices must be integers or slices, not str" when trying to make a field nullable

### ISSUE THREAD

**Issue by eino** [fastavro/fastavro#406]
Hi,

I'm using fastavro to deserialize data ; however I'm having an issue when trying to handle a schema migration that makes a field nullable. As I still need to accept the old data, I followed the Schema resolution rules, especially the following one : 

>  if reader's is a union, but writer's is not
> The first schema in the reader's union that matches the writer's schema is recursively resolved against it. If none match, an error is signalled.

So the new reader's schema will have an union containing the existing record and null. As I don't need to handle a default case, they can be in either order. 
As I understand the specification, the best bet is to put the matching record first in the union, so they match first. 

However deserialization generates the following TypeError : 

> Traceback (most recent call last):
>   File "test.py", line 30, in <module>
>     for record in avro_reader:
>   File "fastavro/_read.pyx", line 675, in _iter_avro_records
>   File "fastavro/_read.pyx", line 557, in fastavro._read._read_data
>   File "fastavro/_read.pyx", line 468, in fastavro._read.read_record
>   File "fastavro/_read.pyx", line 557, in fastavro._read._read_data
>   File "fastavro/_read.pyx", line 457, in fastavro._read.read_record
> TypeError: list indices must be integers or slices, not str

Here are schemas that reproduce the issue : 
Writer schema : 
```
{
    "namespace": "com.eino",
    "name": "Item",
    "type": "record",
    "fields": [
        {"name": "category",
            "type": {
                "type": "record",
                "name":"Category",
                "fields": [
                    {"name": "name", "type": "string"}                    
                ]
            }
        }
    ]
}

```
Reader schema : 

```
{
    "namespace": "com.eino",
    "name": "Item",
    "type": "record",
    "fields": [
        {"name": "category",
            "type": [
                {
                    "type": "record",
                    "name":"Category",
                    "fields": [
                        {"name": "name", "type": "string"}
                    ]
                },
                "null"
            ]
        }
    ]
}
```

And a sample code

```
import avro
from fastavro import reader, parse_schema
from avro.datafile import DataFileWriter
from avro.io import DatumWriter
import json


data = {
    "category": {
        "name": "my-category"
    }
}

schema_file = open("writerschema.avsc", "r")
                                    
orders_schema = avro.schema.Parse(schema_file.read())

out_file = open("data.avro", "wb")
writer = DataFileWriter(out_file, DatumWriter(), orders_schema)
writer.append(data)
writer.close()

with open("readerschema.avsc", "r") as f:
        reader_schema_dict = json.load(f)
reader_schema = parse_schema(reader_schema_dict)
with open("data.avro", "rb") as avro_stream:
    avro_reader = reader(avro_stream, reader_schema=reader_schema)
    for record in avro_reader:
        print(record)
```

If I write the schema with the reader's schema, everything works. So It seems the schema itself is correct.

Any Idea ? Shouldn't fastavro deserialization work with the previous example ?
I'm using the last 0.22.13 by the way.

Thanks in advance !

**Comment by eino** [comment]
Also another issue is, "null" is first in the union, the following error is generated :

> fastavro._read_common.SchemaResolutionError: Schema mismatch: {'type': 'record', 'name': 'com.eino.Category', 'fields': [{'name': 'name', 'type': 'string'}]} is not null

So it seems only the fist item is matched.

When investigating both issues I found that the case where the reader was a union (and the reader was not) was not fully functional ; I hope you don't mind I opened a pull request (#407 ) to fix them ? It works for my sample with those modifications. I hope my changes can help improve this library.

Thanks
